python/http_ece/__init__.py

```diff
@@ -86,47 +153,99 @@ def lengthPrefix(key):
         info=nonceinfo,
         backend=default_backend()
     )
-    result = (hkdf_key.derive(secret), hkdf_nonce.derive(secret))
-    return result
+    try:
+        return hkdf_key.derive(secret), hkdf_nonce.derive(secret)
+    except TypeError as ex:
+        print (ex)
+
 
 def iv(base, counter):
     if (counter >> 64) != 0:
-        raise Exception(u"Counter too big")
+        raise ECEException(u"Counter too big")
     (mask,) = struct.unpack("!Q", base[4:])
     return base[:4] + struct.pack("!Q", counter ^ mask)
 
-def decrypt(buffer, salt, key=None, keyid=None, dh=None, rs=4096, authSecret=None, padSize=2):
-    def decryptRecord(key, nonce, counter, buffer):
+
+def decrypt(content, salt, key=None, keyid=None, dh=None, rs=4096,
+            auth_secret=None, pad_size=2, content_type="aesgcm", **kwargs):
+
+    def parse_content_header(content):
+        """Parse an aes128gcm content body and extract the header values.
+
+        :param content: The encrypted body of the message
+        :type content: bytearray
+
+        """
+        id_len = struct.unpack("!B", content[20])[0]
+        reply = {
+            "salt": content[:16],  # 16 octets
+            "rs": struct.unpack("!l", content[16:20])[0],  # 4 octets
+            "id_len": id_len,  # 1 octet
+            "key_id": content[21:21 + id_len],  # n octets
+            "content": content[21 + id_len:],  # remainder is content
+        }
+        return reply
+
+    def decrypt_record(key, nonce, counter, content):
         decryptor = Cipher(
             algorithms.AES(key),
-            modes.GCM(iv(nonce, counter), tag=buffer[-16:]),
+            modes.GCM(iv(nonce, counter), tag=content[-16:]),
             backend=default_backend()
         ).decryptor()
-        data = decryptor.update(buffer[:-16]) + decryptor.finalize()
-        pad = functools.reduce(lambda x, y: x << 8 | y, struct.unpack("!" + ("B" * padSize), data[0:padSize]))
-        if data[padSize:padSize+pad] != (b"\x00" * pad):
-            raise Exception(u"Bad padding")
-        data = data[padSize+pad:]
+        data = decryptor.update(content[:-16]) + decryptor.finalize()
+        pad = functools.reduce(
+            lambda x, y: x << 8 | y, struct.unpack(
+                "!" + ("B" * pad_size), data[0:pad_size])
+        )
+        if data[pad_size:pad_size+pad] != (b"\x00" * pad):
+            raise ECEException(u"Bad padding")
+        data = data[pad_size + pad:]
         return data
 
-    (key_, nonce_) = deriveKey(mode="decrypt", salt=salt,
-                               key=key, keyid=keyid, dh=dh,
-                               authSecret=authSecret, padSize=padSize)
-    if rs <= padSize:
-        raise Exception(u"Record size too small")
-    rs += 16 # account for tags
-    if len(buffer) % rs == 0:
-        raise Exception(u"Message truncated")
+    # handle old, malformed args
+    pad_size = kwargs.get('padSize', pad_size)
+    auth_secret = kwargs.get('authSecret', auth_secret)
```

Comment:
I don't think that you need to include this.  I would prefer to leave code that hasn't been updated using the old code.  To that end, pad_size can be removed entirely.